File: python/442.find_all_duplicates_in_array.py
from typing import List


class Solution:
    def findDuplicates(self, nums: List[int]) -> List[int]:

        res = []

        for num in nums:
            index = abs(num) - 1
            if nums[index] < 0:
                res.append(index + 1)
            nums[index] *= -1
        return res

File: cpp/442.find_all_duplicates_in_array.cpp
#include <cmath>
#include <vector>

class Solution {
 public:
    std::vector<int> findDuplicates(std::vector<int>& nums) {
        std::vector<int> result;

        for (auto num : nums) {
            // [2, 5, 2, 1, 1, 4]
            // idx = 2 - 1 = 1
            // nums[1] = 5 * -1 = -5
            int idx = std::abs(num) - 1;
            if (nums[idx] < 0) {
                result.push_back(idx + 1);
            }
            nums[idx] *= -1;
        }
        return result;
    }
};
